data-prepper-plugins/kinesis-source/src/main/java/org/opensearch/dataprepper/plugins/kinesis/source/KinesisService.java

```diff
@@ -0,0 +1,169 @@
+    private static final int GRACEFUL_SHUTDOWN_WAIT_INTERVAL_SECONDS = 20;
+
+    private final PluginMetrics pluginMetrics;
+    private final PluginFactory pluginFactory;
+
+    private final String applicationName;
+    private final String tableName;
+    private final String kclMetricsNamespaceName;
+    private final String pipelineName;
+    private final AcknowledgementSetManager acknowledgementSetManager;
+    private final KinesisSourceConfig kinesisSourceConfig;
+    private final KinesisAsyncClient kinesisClient;
+    private final DynamoDbAsyncClient dynamoDbClient;
+    private final CloudWatchAsyncClient cloudWatchClient;
+    private final WorkerIdentifierGenerator workerIdentifierGenerator;
+    private final InputCodec codec;
+
+    @Setter
+    private Scheduler scheduler;
+    private final ExecutorService executorService;
+
+    public KinesisService(final KinesisSourceConfig kinesisSourceConfig,
+                          final KinesisClientFactory kinesisClientFactory,
+                          final PluginMetrics pluginMetrics,
+                          final PluginFactory pluginFactory,
+                          final PipelineDescription pipelineDescription,
+                          final AcknowledgementSetManager acknowledgementSetManager,
+                          final KinesisLeaseConfigSupplier kinesisLeaseConfigSupplier,
+                          final WorkerIdentifierGenerator workerIdentifierGenerator
+                          ){
+        this.kinesisSourceConfig = kinesisSourceConfig;
+        this.pluginMetrics = pluginMetrics;
+        this.pluginFactory = pluginFactory;
+        this.acknowledgementSetManager = acknowledgementSetManager;
+        if (kinesisLeaseConfigSupplier.getKinesisExtensionLeaseConfig().isEmpty()) {
+            throw new IllegalStateException("Lease Coordination table should be provided!");
+        }
+        KinesisLeaseConfig kinesisLeaseConfig =
+                kinesisLeaseConfigSupplier.getKinesisExtensionLeaseConfig().get();
+        this.tableName = kinesisLeaseConfig.getLeaseCoordinationTable().getTableName();
+        this.kclMetricsNamespaceName = this.tableName;
+        this.dynamoDbClient = kinesisClientFactory.buildDynamoDBClient(kinesisLeaseConfig.getLeaseCoordinationTable().getAwsRegion());
+        this.kinesisClient = kinesisClientFactory.buildKinesisAsyncClient(kinesisSourceConfig.getAwsAuthenticationConfig().getAwsRegion());
+        this.cloudWatchClient = kinesisClientFactory.buildCloudWatchAsyncClient(kinesisLeaseConfig.getLeaseCoordinationTable().getAwsRegion());
+        this.pipelineName = pipelineDescription.getPipelineName();
+        this.applicationName = pipelineName;
+        this.workerIdentifierGenerator = workerIdentifierGenerator;
+        this.executorService = Executors.newFixedThreadPool(1);
+        final PluginModel codecConfiguration = kinesisSourceConfig.getCodec();
```

Comment:
Is this codec the type of data they send to Kinesis? It's not just JSON records from Kinesis streams? Or is it always provided from Kinesis as just bytes?

Reply:
When we read from Kinesis, it always comes in as bytes. [Reference Record](https://docs.aws.amazon.com/kinesis/latest/APIReference/API_Record.html#API_Record_Contents)